sources/vega/snapshot/v1/snapshot.proto

```diff
@@ -4,77 +4,271 @@ package vega.snapshot.v1;
 option go_package = "code.vegaprotocol.io/protos/vega/snapshot/v1";
 
 import "vega/vega.proto";
+import "vega/markets.proto";
 import "vega/assets.proto";
+import "vega/chain_events.proto";
 import "vega/governance.proto";
 
-// Snapshot is the entire checkpoint serialised (basically serialised the Checkpoint message + hash)
+enum Format {
+  // Default value, always invalid
+  FORMAT_UNSPECIFIED = 0;
+  // Standard proto encoding
+  FORMAT_PROTO = 1;
+  // Compressed proto, might come in handy
+  FORMAT_PROTO_COMPRESSED = 2;
+  // Simple JSON serialisation, probably not going to be used
+  FORMAT_JSON = 3;
+}
+
+// Represents a set of time intervals that are used when querying for candle-stick data
 message Snapshot {
-  bytes hash = 1;
-  bytes state = 2;
+  uint64 height = 1;
+  Format format = 2;
+  uint32 chunks = 3;
+  bytes hash = 4;
+  bytes metadata = 5;
 }
 
-// Checkpoint aggregates the various engine snapshots
-message Checkpoint {
-  bytes governance = 1;
-  bytes assets = 2;
-  bytes collateral = 3;
-  bytes network_parameters = 4;
-  bytes delegation = 5;
-  bytes epoch = 6; // will just be an epoch event
-  bytes block = 7;
+message NodeHash {
+  string full_key = 1; // namespace<separator>key
+  string namespace = 2;
+  string key = 3;
+  string hash = 4;
+}
+
+message Metadata {
+  int64 version = 1;
+  repeated string chunk_hashes = 2;
+  repeated NodeHash node_hashes = 3; // hashes used by snapshot engine, different to chunk hash if data was split up, or combined into a single chunk
+}
+
+// Chunk is simply an as-is chunk belonging to a snapshot
+// not sure how usable this type would be
+message Chunk {
```

Comment:
what is it used for though? as in chunks must be of size less than 16MB, how can you enforce it this way where you don't control the size of the byte array that the chunk gets serialised into? I'm probably missing something here.

Reply:
This type will be used to marshal/unmarshal an arbitrary number of the Payload messages. Combined, they will form a chunk of the desired size.

Ensuring we don't exceed the 16MB limit is going to be handled in the snapshot engine. For example:

```go
const maxSize = 16000000 // 16MB - ignoring the 1024 factor stuff

markets := e.exec.Snapshot("markets") // returns []ExecutionMarkets
payload := Payload{
    Data: markets,
}
all, _ := proto.Marshal(payload)
totalBytes := len(all)
chunks := totalBytes / maxSize
if i := chunks * maxSize; i < totalBytes {
    chunks++
}
// divide the market data into several chunks
payloads := make([]Payload, 0, chunks)
offset := len(markets) / chunks
for i := 0; i < len(markets); i += offset {
   end := i + offset
   if end > len(markets) {
       end = len(markets)
   }
   payloads = append(payloads, Payload{
       Data: markets[i:end],
   })
}
// serialise each payload as a separate chunk
```

Likewise if we have room to spare, we will combine things like the current epoch, a general AppState (block hash, height, vegatime), and the next checkpoint time into a single chunk. That's why this type is basically a slice of the `one of` wrapper. When we receive a chunk, we can unmarshal it, iterate over the data slice, and dispatch the data to the engines accordingly (or, in case data was split over multiple chunks, we can wait until we've received all data first, then pass it on). To accomplish the latter, I'll need to add some more fields to the Payload message type (so I know how many chunks I need to receive before I have a complete dataset)